src/Streamly/Internal/Data/Stream/IsStream/Top.hs

```diff
@@ -505,12 +580,22 @@ intersectBy eq s1 s2 =
 --
 -- Time: O(m+n)
 --
--- /Unimplemented/
+-- /Pre-release/
 {-# INLINE mergeIntersectBy #-}
-mergeIntersectBy :: -- (IsStream t, Monad m) =>
-    (a -> a -> Ordering) -> t m a -> t m a -> t m a
-mergeIntersectBy _eq _s1 _s2 = undefined
-
+mergeIntersectBy :: (IsStream t, Monad m, Monad (t m), Ord a) =>
+    (a -> a -> Bool) -> t m a -> t m a -> t m a
+mergeIntersectBy eq s1 s2 = concatM $ do
+    -- This may work well when s2 is small
+    let sq = Stream.uniqBy eq s2
+    let k1 = do
+            a <- s1
+            let ss =  Stream.filter (eq a)
+                    $ Stream.takeWhile (a >=) sq
```

Comment:
Same discussion as in mergeJoins applies here as well. The complexity of the algo should be O(m+n). All merge style operations should have linear complexity as we do not need to go through the streams twice. You can take two sorted lists and see how the algo works, you just need to look at the current elements of the two streams to determine what to do.